src/codec/lz4/encoder.rs

```diff
@@ -0,0 +1,335 @@
+                    if let Some(buffer) = self.maybe_buffer.as_mut() {
+                        output.copy_unwritten_from(buffer);
+                    }
+
+                    if self
+                        .maybe_buffer
+                        .as_ref()
+                        .is_none_or(|buffer| buffer.unwritten().is_empty())
+                    {
+                        let len = self.write(Lz4Fn::Flush::<&[u8]>, output)?;
+                        len == 0
+                    } else {
+                        false
+                    }
+                }
+
+                State::Footer => {
+                    if let Some(buffer) = self.maybe_buffer.as_mut() {
+                        output.copy_unwritten_from(buffer);
+                    }
+
+                    if self
+                        .maybe_buffer
+                        .as_ref()
+                        .is_none_or(|buffer| buffer.unwritten().is_empty())
+                    {
+                        self.state = State::Done;
+                        true
+                    } else {
+                        false
+                    }
+                }
+
+                State::Done => true,
+            };
+
+            if done {
+                return Ok(true);
+            }
+
+            if output.unwritten().is_empty() {
+                return Ok(false);
+            }
+        }
+    }
+
+    fn finish(
+        &mut self,
+        output: &mut PartialBuffer<impl AsRef<[u8]> + AsMut<[u8]>>,
+    ) -> Result<bool> {
+        loop {
+            match self.state {
+                State::Header => {
+                    self.write(Lz4Fn::Begin::<&[u8]>, output)?;
+                }
+
+                State::Encoding => {
+                    if let Some(buffer) = self.maybe_buffer.as_mut() {
+                        output.copy_unwritten_from(buffer);
+                    }
+
+                    if self
+                        .maybe_buffer
+                        .as_ref()
+                        .is_none_or(|buffer| buffer.unwritten().is_empty())
+                    {
+                        self.write(Lz4Fn::End::<&[u8]>, output)?;
+                    }
+                }
+
+                State::Footer => {
+                    if let Some(buffer) = self.maybe_buffer.as_mut() {
+                        output.copy_unwritten_from(buffer);
+                    }
+
+                    if self
+                        .maybe_buffer
+                        .as_ref()
+                        .is_none_or(|buffer| buffer.unwritten().is_empty())
+                    {
+                        self.state = State::Done;
+                    }
+                }
+
+                State::Done => {}
+            }
+
+            if let State::Done = self.state {
+                return Ok(true);
+            }
+
+            if output.unwritten().is_empty() {
+                return Ok(false);
+            }
+        }
```

Comment:
```suggestion
        self.flush(output)
```

Reply:
`flush` is similar, but they differ in the function call in the `State::Encoding` case and when they're considered done.